### PR TITLE
feat: add locale-specific quote styles and QuoteStyle value class

### LINKED ISSUE #30: [Feature] Locale-specific Quote Styles
## Problem Statement

Currently, `SmartyPantsLocale` is defined with five locale presets (`en`, `ko`, `ja`, `zhHant`, `zhHans`) and the `SmartyPantsConfig.locale` field stores the user's choice. However, the quote transformation in `_applyTransformations()` **completely ignores the locale** and always produces English-style curly quotes (`U+201C` / `U+201D`) regardless of which locale is configured.

The relevant code in `lib/src/smartypants_base.dart` (lines 350-356):

```dart
if (config.quotes) {
  output = output.replaceAllMapped(
    _quotePattern,
    (match) => '\u201C${match[1]}\u201D',
  );
  output = output.replaceAll("'", '\u2019');
}
```

The hardcoded `\u201C` and `\u201D` values mean that every locale receives identical English-style quotes. This is typographically incorrect for most non-English languages.

### Expected vs. Actual behavior by locale

| Locale | Expected open/close | Currently produced |
|--------|--------------------|--------------------|
| en     | " " (U+201C, U+201D) | " " (correct) |
| fr     | « » (U+00AB, U+00BB) | " " (wrong) |
| de     | „ " (U+201E, U+201C) | " " (wrong) |
| ko/ja  | 「 」(U+300C, U+300D) | " " (wrong) |
| zhHant | 「 」/ 『 』 | " " (wrong) |
| zhHans | " " (U+201C, U+201D) | " " (correct by coincidence) |

Additionally, there is no support for French (`fr`) or German (`de`) locales in the `SmartyPantsLocale` enum, even though these languages have distinct typographic quoting conventions that are widely used.

The `SmartyPantsLocale` enum currently exists as a no-op for quote handling, giving users a false sense of locale-awareness.

### 1. Add `QuoteStyle` value class to `lib/src/smartypants_base.dart`

Introduce an immutable value class that encapsulates primary and secondary (nested) quote characters for a given style:

```dart
class QuoteStyle {
  final String open;
  final String close;
  final String? secondaryOpen;
  final String? secondaryClose;

  const QuoteStyle({
    required this.open,
    required this.close,
    this.secondaryOpen,
    this.secondaryClose,
  });

  static const QuoteStyle english = QuoteStyle(
    open: '\u201C', close: '\u201D',
    secondaryOpen: '\u2018', secondaryClose: '\u2019',
  );
  static const QuoteStyle french = QuoteStyle(
    open: '\u00AB', close: '\u00BB',
    secondaryOpen: '\u2039', secondaryClose: '\u203A',
  );
  static const QuoteStyle german = QuoteStyle(
    open: '\u201E', close: '\u201C',
    secondaryOpen: '\u201A', secondaryClose: '\u2018',
  );
  static const QuoteStyle cjkCornerBracket = QuoteStyle(
    open: '\u300C', close: '\u300D',
    secondaryOpen: '\u300E', secondaryClose: '\u300F',
  );
}
```

### 2. Add `fr` and `de` to `SmartyPantsLocale` enum

```dart
enum SmartyPantsLocale {
  en,
  ko,
  ja,
  zhHant,
  zhHans,
  fr,
  de,
}
```

### 3. Add `customQuoteStyle` field to `SmartyPantsConfig`

```dart
class SmartyPantsConfig {
  final QuoteStyle? customQuoteStyle;

  const SmartyPantsConfig({
    this.customQuoteStyle,
  });
}
```

### 4. Implement `_quoteStyleForLocale()` resolver

```dart
static QuoteStyle _quoteStyleForLocale(SmartyPantsLocale locale) {
  switch (locale) {
    case SmartyPantsLocale.fr:
      return QuoteStyle.french;
    case SmartyPantsLocale.de:
      return QuoteStyle.german;
    case SmartyPantsLocale.ko:
    case SmartyPantsLocale.ja:
    case SmartyPantsLocale.zhHant:
      return QuoteStyle.cjkCornerBracket;
    case SmartyPantsLocale.zhHans:
    case SmartyPantsLocale.en:
    default:
      return QuoteStyle.english;
  }
}
```

### 5. Modify `_applyTransformations()` to use locale-resolved quotes

```dart
if (config.quotes) {
  final style = config.customQuoteStyle
      ?? _quoteStyleForLocale(config.locale);
  output = output.replaceAllMapped(
    _quotePattern,
    (match) => '${style.open}${match[1]}${style.close}',
  );
  output = output.replaceAll("'", '\u2019');
}
```

### 6. Update `copyWith()` for `customQuoteStyle`

```dart
SmartyPantsConfig copyWith({
  QuoteStyle? customQuoteStyle,
}) {
  return SmartyPantsConfig(
    customQuoteStyle: customQuoteStyle ?? this.customQuoteStyle,
  );
}
```

### Tasks

- [ ] Define `QuoteStyle` value class with `open`, `close`, `secondaryOpen`, `secondaryClose` fields in `lib/src/smartypants_base.dart`
- [ ] Add static `QuoteStyle` constants: `english`, `french`, `german`, `cjkCornerBracket`
- [ ] Add `fr` (French) and `de` (German) values to `SmartyPantsLocale` enum with documentation comments
- [ ] Add `customQuoteStyle` field to `SmartyPantsConfig` constructor
- [ ] Update `SmartyPantsConfig.copyWith()` to support `customQuoteStyle`
- [ ] Implement `_quoteStyleForLocale()` static method in `SmartyPants`
- [ ] Modify `_applyTransformations()` to resolve and use `QuoteStyle` instead of hardcoded `\u201C`/`\u201D`
- [ ] Add unit tests for French locale (`fr`) producing `«`/`»` quotes
- [ ] Add unit tests for German locale (`de`) producing `„`/`"` quotes
- [ ] Add unit tests for Korean/Japanese locales producing `「`/`」` quotes
- [ ] Add unit tests for `zhHant` producing CJK corner brackets and `zhHans` producing English-style quotes
- [ ] Add unit tests for `customQuoteStyle` overriding locale defaults
- [ ] Add unit tests verifying backward compatibility: default `en` locale produces identical output to current behavior
- [ ] Update example app to demonstrate locale-specific quote transformations
- [ ] Update `SmartyPantsConfig` dartdoc to document quote style resolution order (`customQuoteStyle` > locale default)

### 1. Locale-specific RegExp patterns instead of `QuoteStyle` value class

Instead of a reusable `QuoteStyle` class, define separate `RegExp` replacement logic per locale.

**Rejected because:** This leads to duplicated transformation logic across locales and makes it impossible for users to supply custom quote styles. The value class approach cleanly separates data (quote characters) from behavior (pattern matching and replacement).

### 2. `SmartyPantsConfig` subclasses per locale

**Rejected because:** This violates the current design where `SmartyPantsConfig` is a flat, `const`-constructible value object. Subclassing prevents `const` construction and complicates the `copyWith()` pattern.

### 3. Remove `SmartyPantsLocale` entirely, use only `customQuoteStyle`

**Rejected because:** The `SmartyPantsLocale` enum already exists in the public API and provides a convenient, discoverable entry point for common locales. Most users should not need to know the exact Unicode code points for their language's quotation marks.

## Expected Impact

- **Backward compatibility preserved**: The default `en` locale continues to produce `\u201C`/`\u201D` quotes.
- **`SmartyPantsLocale` becomes functional for quotes**: The locale field, which currently has no effect on quote output, will drive correct typographic behavior.
- **Broader language coverage**: Adding `fr` and `de` locales brings support for two of the most distinct Western European quoting conventions.
- **CJK quote correctness**: Korean, Japanese, and Traditional Chinese text will receive culturally appropriate corner bracket quotation marks.
- **Extensibility via `customQuoteStyle`**: Users with uncommon locale requirements can supply their own `QuoteStyle` without waiting for enum additions.
- **Minimal API surface growth**: Only one new class (`QuoteStyle`), one new config field (`customQuoteStyle`), and two new enum values (`fr`, `de`) are added.

### PR DESCRIPTION
**Pull request by eunice-hong**
## Summary

Introduce locale-aware quote transformation so that `SmartyPantsLocale` actually drives the output quote characters. Previously, every locale produced hardcoded English curly quotes regardless of configuration.

## Background / Motivation

`SmartyPants._applyTransformations()` hardcoded `\u201C`/`\u201D` for all locales, making `SmartyPantsLocale` a no-op for quote output and giving users a false sense of locale-awareness. French, German, Korean, Japanese, and Traditional Chinese users all received typographically incorrect English-style quotes.

## Related Issues

- Closes #30

## Changes

- **`QuoteStyle` value class** — immutable, `const`-constructible class encapsulating `open`, `close`, `secondaryOpen`, `secondaryClose` characters; includes four static presets: `english`, `french`, `german`, `cjkCornerBracket`
- **`SmartyPantsLocale` enum** — added `fr` (French, guillemets `«»`) and `de` (German, low-high quotes `„"`)
- **`SmartyPantsConfig.customQuoteStyle`** — new optional field; when non-null, overrides the locale default (resolution order: `customQuoteStyle` > locale default)
- **`SmartyPantsConfig.copyWith()`** — updated to support `customQuoteStyle`
- **`SmartyPants._quoteStyleForLocale()`** — new private method that maps each `SmartyPantsLocale` value to a `QuoteStyle`
- **`SmartyPants._applyTransformations()`** — replaced hardcoded `\u201C`/`\u201D` with the resolved `QuoteStyle`

Locale → quote style mapping:

| Locale | Quote style |
|--------|-------------|
| `en`, `zhHans` | `"` / `"` (U+201C, U+201D) |
| `fr` | `«` / `»` (U+00AB, U+00BB) |
| `de` | `„` / `"` (U+201E, U+201C) |
| `ko`, `ja`, `zhHant` | `「` / `」` (U+300C, U+300D) |

## Test Plan

- Commands run:
    - `dart test`
- Notes:
    - Added `'Locale-specific quote styles'` test group in `test/smartypants_test.dart` covering all seven locales, `customQuoteStyle` override, arbitrary custom characters, and `copyWith()` propagation (10 new tests)
    - All 197 tests pass; no existing tests modified

## Documentation (If Needed)

- Dartdoc added to `QuoteStyle` class and all its members
- `SmartyPantsConfig.customQuoteStyle` field documented with resolution-order note
- `SmartyPantsLocale.fr` and `.de` enum values documented
- Example app updated: new `'Locale Quotes'` category demonstrates `fr`, `de`, `ko`, `ja`, and a custom `[`/`]` style

## Breaking Change (If Any)

- Migration notes:
    - None. Default `en` locale continues to produce `\u201C`/`\u201D`. All existing public API signatures are additive-only.

### CODE REVIEW

File: lib/src/smartypants_base.dart [hunk 1]
```diff
@@ -171,6 +256,7 @@ class SmartyPantsConfig {
       cjkEllipsisNormalization:
           cjkEllipsisNormalization ?? this.cjkEllipsisNormalization,
       cjkAngleBrackets: cjkAngleBrackets ?? this.cjkAngleBrackets,
+      customQuoteStyle: customQuoteStyle ?? this.customQuoteStyle,
```
[the remark this answers is not in the document]
Comment (eunice-hong) [reply lib/src/smartypants_base.dart]: Resolved by f2b2d9917d0efb79cd8766db64c3010eda04ff78

**Comment by eunice-hong** [comment]
@codex

File: lib/src/smartypants_base.dart [hunk 2]
```diff
@@ -157,6 +246,7 @@ class SmartyPantsConfig {
     bool? whitespaceNormalization,
     bool? cjkEllipsisNormalization,
     bool? cjkAngleBrackets,
+    Object? customQuoteStyle = _kUnset,
```
[the remark this answers is not in the document]
Comment (eunice-hong) [reply lib/src/smartypants_base.dart]: Resolved by a3748c45d90e28dfa52fa655a5570e65dc058b16

File: lib/src/smartypants_base.dart [hunk 3]
```diff
@@ -348,9 +457,11 @@ class SmartyPants {
...
 
     if (config.quotes) {
+      final style =
+          config.customQuoteStyle ?? _quoteStyleForLocale(config.locale);
       output = output.replaceAllMapped(
         _quotePattern,
-        (match) => '\u201C${match[1]}\u201D',
+        (match) => '${style.open}${match[1]}${style.close}',
       );
       output = output.replaceAll("'", '\u2019');
```
[the remark this answers is not in the document]
Comment (eunice-hong) [reply lib/src/smartypants_base.dart]: Resolved by aeac71f751cf7cb7d4350f940cd1ee5bf982ac9b

**Comment by eunice-hong** [comment]
@codex

File: lib/src/smartypants_base.dart [hunk 4]
```diff
@@ -348,11 +451,13 @@ class SmartyPants {
...
 
     if (config.quotes) {
+      final style =
+          config.customQuoteStyle ?? _quoteStyleForLocale(config.locale);
+      output = output.replaceAll("'", '\u2019');
       output = output.replaceAllMapped(
         _quotePattern,
-        (match) => '\u201C${match[1]}\u201D',
+        (match) => '${style.open}${match[1]}${style.close}',
       );
```
[the remark this answers is not in the document]
Comment (eunice-hong) [reply lib/src/smartypants_base.dart]: Resolved by e2b4cd1e979ef439b573948a29f2a34e19379d31

**Comment by eunice-hong** [comment]
@codex

File: lib/src/smartypants_base.dart [hunk 5]
```diff
@@ -383,6 +478,19 @@ class SmartyPants {
...
+    // Quotes are applied last so custom delimiters are not rewritten by
+    // subsequent passes (e.g. a delimiter containing "--" would otherwise
+    // be converted to an en dash by the dashes pass above).
+    if (config.quotes) {
+      final style =
+          config.customQuoteStyle ?? _quoteStyleForLocale(config.locale);
+      output = output.replaceAll("'", '\u2019');
+      output = output.replaceAllMapped(
+        _quotePattern,
+        (match) => '${style.open}${match[1]}${style.close}',
```
[the remark this answers is not in the document]
Comment (eunice-hong) [reply lib/src/smartypants_base.dart]: Resolved by 4ee68284a0bf89d77142a7477410f9af73a22336

**Comment by eunice-hong** [comment]
@codex

File: lib/src/smartypants_base.dart [hunk 6]
```diff
@@ -383,6 +507,23 @@ class SmartyPants {
...
+    // subsequent passes (e.g. a delimiter containing "--" would otherwise
+    // be converted to an en dash by the dashes pass above).
+    if (config.quotes) {
+      final style =
+          config.customQuoteStyle ?? _quoteStyleForLocale(config.locale);
+      final open =
+          _escapeSentinels(style.open, doubleAngleMarker, markerEscape);
+      final close =
+          _escapeSentinels(style.close, doubleAngleMarker, markerEscape);
+      output = output.replaceAll("'", '\u2019');
```
[the remark this answers is not in the document]
Comment (eunice-hong) [reply lib/src/smartypants_base.dart:649]: Resolved by b86c89ffdf521b7763ec4588ec69ba4d3083d3d9

**Comment by eunice-hong** [comment]
@codex

File: lib/src/smartypants_base.dart [hunk 7]
```diff
@@ -183,8 +270,54 @@ class SmartyPantsConfig {
 /// `<textarea>`) are preserved and never transformed.
 class SmartyPants {
   static final _quotePattern = RegExp(r'"([^"]+)"');
+  static final _singleQuotePattern = RegExp(r"'([^']+)'");
```
[the remark this answers is not in the document]
Comment (eunice-hong) [reply lib/src/smartypants_base.dart]: Resolved by 502497d9f5da2e31bb4dd328b520c660ede472d3

**Comment by eunice-hong** [comment]
@codex

File: lib/src/smartypants_base.dart [hunk 8]
```diff
@@ -183,8 +270,59 @@ class SmartyPantsConfig {
 /// `<textarea>`) are preserved and never transformed.
 class SmartyPants {
   static final _quotePattern = RegExp(r'"([^"]+)"');
+  static final _singleQuotePattern = RegExp(r"'((?:[^']|(?<=\w)'(?=\w))+)'");
```
[the remark this answers is not in the document]
Comment (eunice-hong) [reply lib/src/smartypants_base.dart]: Resolved by 8aa3ffac93e03b8dfddffb24033d0d871555c068

**Comment by eunice-hong** [comment]
@codex

File: lib/src/smartypants_base.dart [hunk 9]
```diff
@@ -183,8 +270,62 @@ class SmartyPantsConfig {
 /// `<textarea>`) are preserved and never transformed.
 class SmartyPants {
   static final _quotePattern = RegExp(r'"([^"]+)"');
+  // Only treat ' as quote boundaries when not between word chars (apostrophes
+  // like in "can't" / "l'homme"); (?<!\w)' = opening, '(?!\w) = closing.
+  static final _singleQuotePattern =
+      RegExp(r"(?<!\w)'((?:[^']|(?<=\w)'(?=\w))*?)'(?!\w)");
```
[the remark this answers is not in the document]
Comment (eunice-hong) [reply lib/src/smartypants_base.dart]: Resolved by 5360c218a23d250563b8033a5f00ccbd1eb40ac5

**Comment by eunice-hong** [comment]
@codex
